File: 2024/3-12-2024/code.lua
File = io.open("input.txt", "r")

Lines = File:read("*all")

Lines = Lines:gsub("%s+", "")

File:close()

-- Part 1 --

function Part1(input)
    local total = 0

    local regex = 'mul%(%d+,%d+%)'

    for match in string.gmatch(input, regex) do
        local x, y = match:match('(%d+),(%d+)')

        local num1 = tonumber(x)
        local num2 = tonumber(y)

        total = total + num1 * num2
    end

    return total
end

function Part2(input)
    local total = 0
    local valid = true
    local pattern = "do%(%)|don't%(%)|mul%(%d+,%d+%)"

    for match in input:gmatch(pattern) do
        if match == "do()" then
            valid = true
        elseif match == "don't()" then
            valid = false
        elseif match:match("mul%(%d+,%d+%)") and valid then
            local x, y = match:match("mul%((%d+),(%d+)%)")
            total = total + tonumber(x) * tonumber(y)
        end
    end
    return total
end

Result1 = Part1(Lines)
print(Result1)

Result2 = Part2(Lines)
print(Result2)

File: 2024/3-12-2024/code.py
import re


def get_input():
    with open("input.txt") as f:
        return f.read().strip()
    

input = get_input()


def part2(input):
    total = 0
    valid = True
    pattern = r"do\(\)|don't\(\)|mul\(\d+,\d+\)"
    matches = re.findall(pattern, input)

    for match in matches :
        if match == "do()":
            valid = True
        elif match == "don't()":
            valid = False
        elif re.match(r'mul\(\d+,\d+\)', match) and valid:
            x = [int(x) for x in re.findall(r'\b\d+\b', match)]
            total += x[0] * x[1]
    return total 


result = part2(input)
print(result)


